### ISSUE TITLE
stack nix integration does not rebuild packages when STACK_IN_NIX_EXTRA_ARGS changes

### ISSUE THREAD

**Issue by bitonic** [commercialhaskell/stack#2802]
I have a setup whereby I have a nix shell with all the C/C++ libraries that I need, and then a Haskell project using them. I'm using `stack --nix` to build the Haskell project.

Now let's say my main Haskell package in the project is A, and it depends on some other Haskell packages such as B, and it also depends on some C libraries that it picks up from nix. Package A has some C files that depend on the C libraries provided by the nix environment.

The way this all works is that the paths to the C libraries are picked up by stack in `STACK_IN_NIX_EXTRA_ARGS`. So the first time I build A, both A and B are going to build with the include dirs and library dirs dictated by `STACK_IN_NIX_EXTRA_ARGS`. So far so good. Note that as a consequence of this both A and B are going to have these include and library dirs in their info, as can be verified with `stack --nix exec -- ghc-pkg describe A` and `stack --nix exec -- ghc-pkg describe B`.

However, say that I upgrade one of the C libraries in the nix environment, thus changing its hash. I then upgrade Haskell package A to use some facility of this new library. What should happen is that _all_ the Haskell packages that were built with the old version of the library are rebuilt, but currently that does not happen -- only A is rebuilt.

The immediate problem that this behavior causes is that when GHC invokes the C compiler to compile the C files in Haskell project A, it provides both the new include dirs (from the latest invocation of ghc by stack) and the old include dirs (remembered by GHC as include dirs in the package info of B). By chance the include dirs in B happen to "win", and thus A cannot see the header files for the new version of the C library.

Besides the symptom described, it seems clear to me that if a C library is change all the Haskell packages that were built with it should be rebuilt too.

### Stack version

```
$ stack --version                                
Version 1.2.0, Git revision 241cd07d576d9c0c0e712e83d947e3dd64541c42 (4054 commits) x86_64 hpack-0.14.0
```

**Comment by mboes** [comment]
Pinging @YPares

**Comment by YPares** [comment]
If both A and B packages are indeed declared in the same stack.yml, it's
indeed not obvious to me why
only A is rebuilt when STACK_IN_NIX_EXTRA_ARGS changes, as this means a
change of command line opts,
and stack normally rebuilds when CLI opts are changed.
More investigation is required...

2016-11-22 15:34 GMT+01:00 Francesco Mazzoli <notifications@github.com>:

**Comment by bitonic** [comment]
Note that in our case package B is an extra dep specified with a git repo
as a location. Not sure of this makes a difference but it's something to
keep in mind.

**Comment by mboes** [comment]
This is a pretty deep problem. I don't think there's an easy way out of it. The issue is that Stack doesn't rebuild snapshot packages when you change your configuration. That's in line with Stack's general model: snapshot packages are built once and for all for each snapshot. We could decide that anytime you change your shell file, the snapshot packages get rebuilt in a different location under `~/.stack/snapshots`. But not only would that be a huge pessimization (Stack would rebuild all snapshot packages all the time), it's also not in line with Stack's "build once" model, so not sure how easy it  would be to implement.

**Comment by bitonic** [comment]
@mboes but `stack` does rebuild a project if the flags change, e.g. when `--extra-lib-dirs` change. Shouldn't this be the same?

**Comment by mboes** [comment]
It rebuilds the local packages, right? Does it rebuild the snapshot packages?

**Comment by bitonic** [comment]
Oops, I didn't read "snapshot" in your messages.

The packages I'm talking about are not snapshot packages, they're packages in the same project. When I change `--extra-lib-dirs`, stack rebuilds them.

**Comment by mboes** [comment]
Ah ok. Much more straightfoward to fix problem then. But do you have a minimal repro? I don't see why you would be seeing the behaviour you say you are seeing. If you change the C package, then `--extra-{lib,include}-dirs` will change, *for every local package*, since they're all built in the same environment. So in principle B would be rebuilt too, even though it doesn't depend on C, and A would be rebuilt because both C and B changed.

**Comment by bitonic** [comment]
That's right, I'd expect them to change. I'll try to get a minimal repro, but for someone familiar with the code it should be easy to verify that the right thing is being done. I'll also try to take a look.

**Comment by YPares** [comment]
@bitonic Hi! Did you get a chance to work on that minimal repro?

**Comment by bitonic** [comment]
@YPares no, not yet.
